apps/HelloiOS/HelloiOS/HalideView.mm

```diff
@@ -0,0 +1,364 @@
+        {0, w, c},
+        {0, h, c * row_stride},
+        {0, c, 1}
+    };
+
+    // This allows us to make a Buffer with an arbitrary shape
+    // and memory managed by Buffer itself
+    pixel_buf = Buffer<uint8_t>(nullptr, 3, pixelBufShape);
+    pixel_buf.allocate();
+}
+
+- (void)didMoveToWindow
+{
+    self.contentScaleFactor = self.window.screen.nativeScale;
+}
+
+- (id)initWithFrame:(CGRect)frame
+{
+    self = [super initWithFrame:frame];
+    if(self)
+    {
+        [self initCommon];
+    }
+    return self;
+}
+
+- (instancetype)initWithCoder:(NSCoder *)coder
+{
+    self = [super initWithCoder:coder];
+    
+    if(self)
+    {
+        [self initCommon];
+    }
+    return self;
+}
+
+- (void)updateLogWith: (double) elapsedTime using_metal: (bool) using_metal
+{
+#if HAS_METAL_SDK
+    const char* mode = using_metal ? "(Metal; Double-tap for CPU)" : "(CPU; Double-tap for Metal)";
+#else
+    const char* mode = "(CPU; Metal not available)";
+#endif
+    char log_text[2048];
+    snprintf(log_text, sizeof(log_text),
+         "Halide routine takes %0.3f ms %s\n", elapsedTime * 1000, mode);
+    [self.outputLog setText: [NSString stringWithUTF8String:log_text]];
+}
+
+- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
+    UITouch* touch = [touches anyObject];
+    self.touch_position = [touch locationInView:self];
+    self.touch_active = [self pointInside:self.touch_position withEvent:event];
+#if HAS_METAL_SDK
+    NSUInteger numTaps = [touch tapCount];
+    if (numTaps > 1) {
+        self.use_metal = !self.use_metal;
+        NSLog(@"TBTaps: %d, self.use_metal %d", (int)numTaps, (int)self.use_metal);
+    }
+#endif
+}
+
+- (void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event {
+    self.touch_position = [touches.anyObject locationInView:self];
+    self.touch_active = [self pointInside:self.touch_position withEvent:event];
+}
+
+- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event {
+    self.touch_active = false;
+}
+
+- (void)touchesCancelled:(NSSet *)touches withEvent:(UIEvent *)event {
+    self.touch_active = false;
+}
+
+- (void)renderOneFrame: (const HalideFuncs &) halide_funcs using_metal: (bool) using_metal
+{
+    int tx = -100, ty = -100;
+    if (self.touch_active) {
+        // Note that buf/bounds is not necessarily the same as self.contentScaleFactor
+        tx = (int) (self.touch_position.x * buf1.dim(0).extent() / self.bounds.size.width);
+        ty = (int) (self.touch_position.y * buf1.dim(1).extent() / self.bounds.size.height);
+        NSLog(@"touch %d %d", tx, ty);
+    }
+
+#if HAS_METAL_SDK
+    const bool output_bgra = true;
+#else
+    const bool output_bgra = false;
+#endif
+    double t_before = CACurrentMediaTime();
+    halide_funcs.update((__bridge void *)self, buf1, tx, ty, iteration, buf2);
+    halide_funcs.render((__bridge void *)self, buf2, output_bgra, pixel_buf);
+    double t_after = CACurrentMediaTime();
```

Comment:
Also occurs to me that timing that is (say) megapixels-per-sec would be more useful.